cpp/dolfinx/fem/Function.h

```diff
@@ -291,32 +304,37 @@ class Function
     interpolate(e, cells);
   }
 
-  /// Evaluate the Function at points
+  /// @brief Evaluate the Function at points.
   ///
   /// @param[in] x The coordinates of the points. It has shape
-  /// (num_points, 3).
+  /// (num_points, 3) and storage is row-major.
+  /// @param[in] xshape The shape of `x`.
   /// @param[in] cells An array of cell indices. cells[i] is the index
   /// of the cell that contains the point x(i). Negative cell indices
   /// can be passed, and the corresponding point will be ignored.
-  /// @param[in,out] u The values at the points. Values are not computed
+  /// @param[out] u The values at the points. Values are not computed
   /// for points with a negative cell index. This argument must be
-  /// passed with the correct size.
-  void eval(const xt::xtensor<double, 2>& x,
-            const std::span<const std::int32_t>& cells,
-            xt::xtensor<T, 2>& u) const
+  /// passed with the correct size. Storage is row-major.
+  /// @param[in] ushape The shape of `u`.
+  void eval(std::span<const double> x, std::array<std::size_t, 2> xshape,
```

Comment:
Could this be `mdspan` here? This C-like interface is too verbose.

Reply:
> Could this be `mdspan` here? This C-like interface is too verbose.

I think, at least for now, it's better to avoid exposing `mdspan` in the public interface where possible. Also, in practice the 'c-like' interface is less verbose because we often have:
```cpp
auto [data, shape] = foo();
bar(data, shape);
```
`mdspan` doesn't help here because it can't be returned by `foo` because `mdspan` doesn't own the data. Similarly the 'c-like' interface is convenient in the pybind layer. In practice, it's rare to see any of this in user code anyway.

Reply:
I agree that for python users this makes no difference, as all interface handling happens in our pybind11 cpp code. For the same reason I think that `mdspan` interface is as convenient as c-like interface, as you only need to construct and pass the mdspan in pybind11 cpp wrappers.

There are many places where
```cpp
std::span<...> x = mesh->geometry().x();
foo(x, {mesh->geometry().x().size() / 3, 3});
```
is really verbose and error prone. I thought that in these places where non-owning data object is returned we'd use `mdspan` instead of the above repeated access to the data to figure out its shape.

Reply:
> I agree that for python users this makes no difference, as all interface handling happens in our pybind11 cpp code. For the same reason I think that `mdspan` interface is as convenient as c-like interface, as you only need to construct and pass the mdspan in pybind11 cpp wrappers.
> 
> There are many places where
> 
> ```c++
> std::span<...> x = mesh->geometry().x();
> foo(x, {mesh->geometry().x().size() / 3, 3});
> ```
> 
> is really verbose and error prone. I thought that in these places where non-owning data object is returned we'd use `mdspan` instead of the above repeated access to the data to figure out its shape.

This is unrelated to the current PR - methods that return a `span` have not been touched.  `xt::adapt` has just been replaced by `mdspan` for wrapping the data. It could be discussed as a future change. Using `mdspan` in these cases would be neater, at the expense of changing an STL type to a non-standard type. `mdspan` does bring some complication over `std::span` - memory is not guaranteed to be contiguous (strides are supported), storage layouts can differ, etc, so `foo.data_handle()` needs to be used with care.